pydantic_ai_slim/pydantic_ai/models/google.py

```diff
@@ -766,20 +776,88 @@ async def _get_event_iterator(self) -> AsyncIterator[ModelResponseStreamEvent]:
                         part=FilePart(content=BinaryContent.narrow_type(content), provider_details=provider_details),
                     )
                 elif part.executable_code is not None:
-                    code_execution_tool_call_id = _utils.generate_tool_call_id()
-                    part = _map_executable_code(part.executable_code, self.provider_name, code_execution_tool_call_id)
-                    part.provider_details = provider_details
-                    yield self._parts_manager.handle_part(vendor_part_id=uuid4(), part=part)
+                    part_obj = self._handle_executable_code_streaming(part.executable_code)
+                    if part_obj is not None:  # pragma: no branch
+                        part_obj.provider_details = provider_details
+                        yield self._parts_manager.handle_part(vendor_part_id=uuid4(), part=part_obj)
                 elif part.code_execution_result is not None:
-                    assert code_execution_tool_call_id is not None
+                    assert self._code_execution_tool_call_id is not None
                     part = _map_code_execution_result(
-                        part.code_execution_result, self.provider_name, code_execution_tool_call_id
+                        part.code_execution_result, self.provider_name, self._code_execution_tool_call_id
                     )
                     part.provider_details = provider_details
                     yield self._parts_manager.handle_part(vendor_part_id=uuid4(), part=part)
                 else:
                     assert part.function_response is not None, f'Unexpected part: {part}'  # pragma: no cover
 
+            # Grounding metadata is attached to the final text chunk, so
+            # we emit the `BuiltinToolReturnPart` after the text delta so
+            # that the delta is properly added to the same `TextPart` as earlier chunks
+            async for event in self._handle_file_search_grounding_metadata_streaming(candidate.grounding_metadata):
+                yield event
+
+    async def _handle_file_search_grounding_metadata_streaming(
+        self, grounding_metadata: GroundingMetadata | None
+    ) -> AsyncIterator[ModelResponseStreamEvent]:
+        """Handle file search grounding metadata for streaming responses.
+
+        Yields a BuiltinToolReturnPart if file search results are available in the grounding metadata.
+        """
+        if not self._file_search_tool_call_id or not grounding_metadata:
+            return
+
+        grounding_chunks = grounding_metadata.grounding_chunks
+        retrieved_contexts = _extract_file_search_retrieved_contexts(grounding_chunks)
+        if retrieved_contexts:  # pragma: no branch
+            yield self._parts_manager.handle_part(
```

Comment:
I'd prefer to keep all the `self._parts_manager.handle_part` calls inside `_get_event_iterator`, so let's change this method to return `BuiltinToolReturnPart | None`